File: Pod/Classes/PURLogStore.m
#import <YapDatabase/YapDatabase.h>
#import "PURLogStore.h"
#import "PURLog.h"
#import "PUROutput.h"

static NSString * const LogDatabaseDirectory = @"com.cookpad.PureeData.default";
static NSString * const LogDatabaseFileName = @"logs.db";
static NSString * const LogDataCollectionNamePrefix = @"log_";

static NSMutableDictionary<NSString *, YapDatabase *> *__databases;

NS_ASSUME_NONNULL_BEGIN

@interface PURLogStore ()

@property (nonatomic) NSString *databasePath;
@property (nonatomic) YapDatabaseConnection *databaseConnection;
@property (nonatomic) dispatch_queue_t databaseReadWriteCompletionQueue;

@end

NS_ASSUME_NONNULL_END

static NSString *PURLogStoreCollectionNameForPattern(NSString *pattern)
{
    return [LogDataCollectionNamePrefix stringByAppendingString:pattern];
}

static NSString *PURLogKey(PUROutput *output, PURLog *log)
{
    return [[NSStringFromClass([output class]) stringByAppendingString:@"_"] stringByAppendingString:log.identifier];
}

@implementation PURLogStore

+ (void)initialize
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        __databases = [NSMutableDictionary new];
    });
}

+ (instancetype)defaultLogStore
{
    return [[self alloc] initWithDatabasePath:[self defaultDatabasePath]];
}

- (instancetype)initWithDatabasePath:(NSString *)databasePath
{
    self = [super init];
    if (self) {
        _databasePath = databasePath;
        _databaseReadWriteCompletionQueue = dispatch_queue_create("PureeLogStoreReadWriteCompletion", NULL);
    }
    return self;
}

- (void)dealloc {
    self.databaseReadWriteCompletionQueue = nil;
}

- (BOOL)prepare
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *databaseDirectory = [self.databasePath stringByDeletingLastPathComponent];
    BOOL isDirectory = NO;
    if (![fileManager fileExistsAtPath:databaseDirectory isDirectory:&isDirectory]) {
        NSError *error = nil;
        [fileManager createDirectoryAtPath:databaseDirectory
               withIntermediateDirectories:YES
                                attributes:nil
                                     error:&error];
        if (error) {
            return NO;
        }
    } else if (!isDirectory) {
        return NO;
    }

    YapDatabase *database = __databases[self.databasePath];
    if (!database) {
        database = [[YapDatabase alloc] initWithPath:self.databasePath];
        __databases[self.databasePath] = database;
    }
    if (self.databaseConnection.database != database) {
        self.databaseConnection = [database newConnection];
    }
    return self.databaseConnection;
}

+ (NSString *)defaultDatabasePath
{
    NSArray *libraryCachePaths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString *libraryCacheDirectoryPath = libraryCachePaths.firstObject;
    NSString *filePath = [LogDatabaseDirectory stringByAppendingPathComponent:LogDatabaseFileName];
    NSString *databasePath = [libraryCacheDirectoryPath stringByAppendingPathComponent:filePath];

    return databasePath;
}

- (void)retrieveLogsForPattern:(NSString *)pattern output:(PUROutput *)output completion:(PURLogStoreRetrieveCompletionBlock)completion;
{
    NSAssert(self.databaseConnection, @"Database connection is not available");

    NSMutableArray<PURLog *> *logs = [NSMutableArray new];
    [self.databaseConnection asyncReadWithBlock:^(YapDatabaseReadTransaction *transaction){
        NSString *keyPrefix = [NSStringFromClass([output class]) stringByAppendingString:@"_"];
        [transaction enumerateRowsInCollection:PURLogStoreCollectionNameForPattern(output.tagPattern)
                                    usingBlock:^(NSString *key, PURLog *log, id metadata, BOOL *stop){
                                        [logs addObject:log];
                                    }
                                    withFilter:^BOOL(NSString *key){
                                        return [key hasPrefix:keyPrefix];
                                    }];
    }
                                completionQueue:self.databaseReadWriteCompletionQueue
                                completionBlock:^{
                                    completion(logs);
                                }];
}

- (void)addLog:(PURLog *)log fromOutput:(PUROutput *)output completion:(nullable dispatch_block_t)completion
{
    NSAssert(self.databaseConnection, @"Database connection is not available");

    [self addLogs:@[ log ] fromOutput:output completion:completion];
}

- (void)addLogs:(NSArray<PURLog *> *)logs fromOutput:(PUROutput *)output completion:(nullable dispatch_block_t)completion
{
    NSAssert(self.databaseConnection, @"Database connection is not available");

    [self.databaseConnection asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction){
        NSString *collectionName = PURLogStoreCollectionNameForPattern(output.tagPattern);
        for (PURLog *log in logs) {
            [transaction setObject:log forKey:PURLogKey(output, log) inCollection:collectionName];
        }
    }
                                     completionQueue:self.databaseReadWriteCompletionQueue
                                     completionBlock:completion];
}

- (void)removeLogs:(NSArray<PURLog *> *)logs fromOutput:(PUROutput *)output completion:(nullable dispatch_block_t)completion
{
    NSAssert(self.databaseConnection, @"Database connection is not available");

    [self.databaseConnection asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction){
        NSString *collectionName = PURLogStoreCollectionNameForPattern(output.tagPattern);
        for (PURLog *log in logs) {
            [transaction removeObjectForKey:PURLogKey(output, log) inCollection:collectionName];
        }
    }
                                     completionQueue:self.databaseReadWriteCompletionQueue
                                     completionBlock:completion];
}

- (void)clearAll
{
    NSAssert(self.databaseConnection, @"Database connection is not available");

    [self.databaseConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction){
        [transaction removeAllObjectsInAllCollections];
    }];
}

@end

File: Pod/Classes/PURLogStore.swift
//
//  PURLogStore.swift
//  Pods
//
//  Created by admin on 7/27/17.
//
//

//  Converted with Swiftify v1.0.6414 - https://objectivec2swift.com/
import Foundation
import YapDatabase

private let LogDatabaseDirectory: String = "com.cookpad.PureeData.default"
private let LogDatabaseFileName: String = "logs.db"
private let LogDataCollectionNamePrefix: String = "log_"
private var __databases = [String: YapDatabase]()

typealias PURLogStoreRetrieveCompletionBlock = (_ logs: [PURLog]) -> Void

private func PURLogStoreCollectionNameForPattern(pattern: String) -> String {
    return LogDataCollectionNamePrefix + (pattern)
}

private func PURLogKey(output: PUROutput, log: PURLog) -> String {
    return String(describing: output.self) + ("_") + (log.identifier)
}

class PURLogStore {
    
    var databasePath: URL?
    var databaseConnection: YapDatabaseConnection?
    
    class func initialize() {
        __databases = [String: YapDatabase]()
    }
    
    init(databasePath: URL?) {
        if databasePath == nil {
            self.databasePath = self.defaultDatabasePath()
        }
        
        self.databasePath = databasePath
    }
    
    func prepare() -> Bool {
        let fileManager = FileManager.default
        let databaseDirectory: String = (databasePath?.deletingLastPathComponent().lastPathComponent)!
        var isDirectory: Bool = false
        
        if !fileManager.fileExists(atPath: databaseDirectory, isDirectory: isDirectory as! UnsafeMutablePointer<ObjCBool>) {
            var error: Error? = nil
            try? fileManager.createDirectory(atPath: databaseDirectory, withIntermediateDirectories: true, attributes: nil)
            if error != nil {
                return false
            }
        } else if !isDirectory {
            return false
        }
        
        var database: YapDatabase? = __databases[(databasePath?.absoluteString)!]
        if database == nil {
            database = YapDatabase(path: (databasePath?.absoluteString)!)
            __databases[(databasePath?.absoluteString)!] = database
        }
        if databaseConnection?.database != database {
            databaseConnection = database?.newConnection()
        }
        return true
    }
    
    private func defaultDatabasePath() -> URL {
        let libraryCachePaths: [Any] = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
        let libraryCacheDirectoryPath: String? = libraryCachePaths.first as! String?
        let filePath: String = URL(fileURLWithPath: LogDatabaseDirectory).appendingPathComponent(LogDatabaseFileName).absoluteString
        let databasePath = URL(fileURLWithPath: libraryCacheDirectoryPath!).appendingPathComponent(filePath)
        
        return databasePath
    }
    
    func retrieveLogs(for output: PUROutput, completion: @escaping PURLogStoreRetrieveCompletionBlock) {
        assert((databaseConnection == nil), "Database connection is not available")
        
        var logs: [PURLog] = [PURLog]()
        
        databaseConnection?.asyncRead({(_ transaction: YapDatabaseReadTransaction) -> Void in
            
            let keyPrefix: String = String(describing: output.self) + ("_")
            
            transaction.enumerateRows(inCollection: PURLogStoreCollectionNameForPattern(pattern: output.tagPattern), using: { (key, log, metadata, stop) in
                logs.append(log as! PURLog)
            }, withFilter: { (key) -> Bool in
                return key.hasPrefix(keyPrefix)
            })
            
        }, completionBlock: {() -> Void in
            completion(logs)
        })
    }
    
    func add(_ log: PURLog, for output: PUROutput, completion: (() -> Void)?) {
        assert((databaseConnection == nil), "Database connection is not available")
        addLogs([log], for: output, completion: completion)
    }
    
    func addLogs(_ logs: [PURLog], for output: PUROutput, completion: (() -> Void)?) {
        assert((databaseConnection == nil), "Database connection is not available")
        databaseConnection?.asyncReadWrite({(_ transaction: YapDatabaseReadWriteTransaction) -> Void in
            let collectionName: String = PURLogStoreCollectionNameForPattern(pattern: output.tagPattern)
            for log: PURLog in logs {
                transaction.setObject(log, forKey: PURLogKey(output: output, log: log), inCollection: collectionName)
            }
        }, completionBlock: completion)
    }
    
    func removeLogs(_ logs: [PURLog], for output: PUROutput, completion: (() -> Void)?) {
        assert((databaseConnection == nil), "Database connection is not available")
        databaseConnection?.asyncReadWrite({(_ transaction: YapDatabaseReadWriteTransaction) -> Void in
            let collectionName: String = PURLogStoreCollectionNameForPattern(pattern: output.tagPattern)
            for log: PURLog in logs {
                transaction.removeObject(forKey: PURLogKey(output: output, log: log), inCollection: collectionName)
            }
        }, completionBlock: completion)
    }
    
    func clearAll() {
        assert((databaseConnection == nil), "Database connection is not available")
        databaseConnection?.readWrite({(_ transaction: YapDatabaseReadWriteTransaction) -> Void in
            transaction.removeAllObjectsInAllCollections()
        })
    }
}


